### PR TITLE
DataSharing communication tests [10213]

### PR DESCRIPTION
**Pull request by jparisu**
Branch over : `feature/DataSharing-pools`

Changes:

1. One test has been added to `test/communication`:
New test using a `FixedSized` type instead of `HelloWorld` type.

2. Change names of `test/dds/communication`:
The only test in this directory was related just with dynamic types.
The files related with this test has been renamed with `Dynamic` addition to names

3. Implement `main` test executables:
Implement `Publisher|Subscriber Main.cpp|Module.hpp|Module.cpp` whit the Publisher and Subscriber implementations and parametrizations needed to check communication.

4. Python test generation:
`test_build.py` automaticly execute a communication test following a test definition given by a JSON file
`definitions_example.json` has an explanation of this JSON file fields.

### CODE REVIEW

**Comment by jparisu** [comment]
New DDS communication tests are not yet working due to an error in QoS or Connectivity.
The Publisher and Subscriber are discovered by each other, and messages are written, but they are not beeing received by the Reader. Thus the test fails.

FUTURE WORK:

- Fix communication error in Publisher and Subscriber implementations.
- Add new .json and .xml files to generate new test cases

**Comment by jparisu** [comment]
Each test implemented is run from a `.json` file with the participants specifications.
Tests implemented:

- `simple_communication.json` - default reliable publisher and default reliable subscriber
- `besteffort_communication.json` -  default best effort publisher and default best effort subscriber
- `mix_communication.json` - default reliable publisher and default reliable subscriber and a default PubSub reliable
- `simple_communication_fixed_type.json` - default reliable publisher and default reliable subscriber with `FixedSizedType` type.
- zero_copy_communication.json` - zero copy publisher and zero copy subscriber
- zero_copy_pub_communication.json` - zero copy publisher and default subscriber (both with `FixedSizedType` type).
- zero_copy_sub_communication.json` - default publisher and zero copy subscriber (both with `FixedSizedType` type).
- mix_zero_copy_communication.json` - default publisher and default subscriber and zero copy publisher and zero copy subscriber (all with `FixedSizedType` type).

File: test/dds/communication/test_build.py [hunk 1]
```diff
@@ -169,16 +169,16 @@ def execute_commands(pub_commands, sub_commands, pubsub_commands, logger):
...
     for pubsub_command in pubsub_commands:
         logger.info(f'Executing pubsub: {pubsub_command}')
         proc = subprocess.Popen(pubsub_command)
         pubsubs_proc.append(proc)
 
+    for pubisher_command in pub_commands:
+        logger.info(f'Executing publisher: {pubisher_command}')
+        proc = subprocess.Popen(pubisher_command)
+        pubs_proc.append(proc)
+
```
Comment (IkerLuengo) [inline test/dds/communication/test_build.py:181]: `pubisher` -> `publisher`

**Review by IkerLuengo — CHANGES_REQUESTED** [review]

File: test/dds/communication/SubscriberModule.cpp [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+bool SubscriberModule::init(
+        uint32_t seed,
+        const std::string& magic)
+{
+    std::cout << "Initializing Subscriber" << std::endl;
+
+    //Do not enable entities on creation
+    // DomainParticipantFactoryQos factory_qos;
+    // factory_qos.entity_factory().autoenable_created_entities = false;
+    // DomainParticipantFactory::get_instance()->set_qos(factory_qos);
```
Comment (MiguelCompany) [inline test/dds/communication/SubscriberModule.cpp:67]: Should we remove this code?

File: test/dds/communication/definitions_example.json [hunk 3]
```diff
@@ -0,0 +1,43 @@
+{
+    "DESCRIPTION" : "This json file specifies all the different parameters each participant class can have",
+    "DESCRIPTION_structure" : "All participants to create must be an element in <participants> array and must have <kind>",
+    "DESCRIPTION_fields" : "Every field must have a value. Booleans : true/false. Strings : string. Int : int",
+    "DESCRIPTION_kind" : "Element <kind> can be 'subscriber' 'publisher' or 'pubsub'",
+    "participants" : [
+        {
+            "kind" : "publisher",
```
Comment (MiguelCompany) [inline test/dds/communication/definitions_example.json:8]: As this is the example, I think it should reflect the default values.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: test/dds/communication/CMakeLists.txt [hunk 4]
```diff
@@ -25,33 +25,118 @@ if(NOT ((MSVC OR MSVC_IDE) AND EPROSIMA_INSTALLER) AND fastcdr_FOUND)
...
+    set(DDS_PUBLISHER_SOURCE ${COMMON_SOURCE}
+        PublisherModule.cpp
+        PublisherMain.cpp
+        )
+    add_executable(DDSCommunicationPublisher ${DDS_PUBLISHER_SOURCE})
+    target_compile_definitions(DDSCommunicationPublisher PRIVATE
+        $<$<AND:$<NOT:$<BOOL:${WIN32}>>,$<STREQUAL:"${CMAKE_BUILD_TYPE}","Debug">>:__DEBUG>
+        $<$<BOOL:${INTERNAL_DEBUG}>:__INTERNALDEBUG> # Internal debug activated.
+        )
+    target_include_directories(DDSCommunicationPublisher PRIVATE ${PROJECT_SOURCE_DIR}/test/blackbox)
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply test/dds/communication/CMakeLists.txt:66]: For the data types headers

File: test/dds/communication/CMakeLists.txt [hunk 5]
```diff
@@ -25,33 +25,118 @@ if(NOT ((MSVC OR MSVC_IDE) AND EPROSIMA_INSTALLER) AND fastcdr_FOUND)
...
         )
-    add_executable(DDSSimpleCommunicationPublisher ${DDS_PUBLISHER_SOURCE})
-    target_compile_definitions(DDSSimpleCommunicationPublisher PRIVATE
+    add_executable(DDSCommunicationSubscriber ${DDS_SUBSCRIBER_SOURCE})
+    target_compile_definitions(DDSCommunicationSubscriber PRIVATE
         $<$<AND:$<NOT:$<BOOL:${WIN32}>>,$<STREQUAL:"${CMAKE_BUILD_TYPE}","Debug">>:__DEBUG>
         $<$<BOOL:${INTERNAL_DEBUG}>:__INTERNALDEBUG> # Internal debug activated.
         )
-    target_link_libraries(DDSSimpleCommunicationPublisher fastrtps fastcdr foonathan_memory ${CMAKE_DL_LIBS})
+    target_include_directories(DDSCommunicationSubscriber PRIVATE ${PROJECT_SOURCE_DIR}/test/blackbox)
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply test/dds/communication/CMakeLists.txt:78]: For the data types headers

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3666/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/846/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3230/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3689/)

File: test/dds/communication/PubSubMain.cpp [hunk 6]
```diff
@@ -147,14 +149,29 @@ int main(
...
+            std::cout << "Wrong argument " << argv[arg_count] << std::endl;
+            return -1;
+        }
 
         ++arg_count;
     }
 
     if (xml_file)
     {
         eprosima::fastrtps::Domain::loadXMLProfilesFile(xml_file);
```
Comment (MiguelCompany) [inline test/dds/communication/PubSubMain.cpp:173]: This line should be removed.

Applies to every DDS application.

File: test/dds/communication/PubSubMain.cpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+            publishers = strtol(argv[arg_count], nullptr, 10);
+        }
+
+        ++arg_count;
+    }
+
+    if (xml_file)
+    {
+        eprosima::fastrtps::Domain::loadXMLProfilesFile(xml_file);
+        //DomainParticipantFactory::get_instance()->load_XML_profiles_file(xml_file);
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/dds/communication/PubSubMain.cpp]: The line was uncommented, but the previous line should be removed, as we are not using `eprosima::fastrtps` namespace

**Review by MiguelCompany — APPROVED** [review]
LGTM 🏅

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3669/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/849/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3232/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3694/)

**Review by MiguelCompany — APPROVED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3673/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/853/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3235/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3698/)

**Comment by MiguelCompany** [comment]
All the tests added/modified by this PR were passing.

**Pull request merged**